src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,165 @@
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = getNodeByIndex(index);
+        T temp = currentNode.value;
+        currentNode.value = value;
+        return temp;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        if (index == 0 && size == 1) {
+            T tempToReturn = head.value;
+            unlinkOneNode();
+            return tempToReturn;
+        } else {
+            Node<T> currentNode = getNodeByIndex(index);
+            remove(currentNode.value);
+            return currentNode.value;
+        }
     }
 
     @Override
-    public boolean remove(T object) {
-        return false;
+    public boolean remove(T value) {
+        if (isEquals(head.value, value) && size == 1) {
+            return unlinkOneNode();
+        } else if (isEquals(head.value, value)) {
+            return unlinkHeadNode();
+        } else {
+            return unlinkNode(value);
+        }
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bound");
+        }
+    }
+
+    private boolean unlinkNode(T value) {
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if (isEquals(currentNode.value, value)) {
+                currentNode.prev.next = currentNode.next;
+                currentNode.next.prev = currentNode.prev;
+                size--;
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
+
+    private Node<T> getNodeByIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bound");
+        }
+        Node<T> currentNode = size / 2 < index ? head : tail;
+        if (currentNode == head) {
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    private boolean unlinkOneNode() {
+        tail.next = null;
+        tail.prev = null;
+        tail.value = null;
+        head = tail;
+        size--;
+        return true;
+    }
+
+    private boolean unlinkHeadNode() {
+        head.next.prev = null;
+        head = head.next;
+        size--;
+        return true;
+    }
```

Comment:
I think these 2 methods can be implemented in your UnlinkNode method. They essentially perform the same function